internal/statemachine/state_machine.go

```diff
@@ -116,6 +127,101 @@ func (stateMachine *StateMachine) validateInput() error {
 	return nil
 }
 
+// parseImageSizes handles the flag --image-size, which is a string in the format
+// <volumeName>:<volumeSize>,<volumeName2>:<volumeSize2>. It can also be in the
+// format <volumeSize> to signify one size to rule them all
+func (stateMachine *StateMachine) parseImageSizes() error {
+	// initialize the size map
+	stateMachine.imageSizes = make(map[string]quantity.Size)
+
+	// If --image-size was not used, simply return
+	if stateMachine.commonFlags.Size == "" {
+		return nil
+	}
+
+	if !strings.Contains(stateMachine.commonFlags.Size, ":") {
+		// handle the "one size to rule them all" case
+		parsedSize, err := quantity.ParseSize(stateMachine.commonFlags.Size)
+		if err != nil {
+			return fmt.Errorf("Failed to parse argument to --image-size: %s", err.Error())
+		}
+		for volumeName := range stateMachine.gadgetInfo.Volumes {
+			stateMachine.imageSizes[volumeName] = parsedSize
+		}
+	} else {
+		allSizes := strings.Split(stateMachine.commonFlags.Size, ",")
+		for _, size := range allSizes {
+			// each of these should be of the form "<name|number>:<size>"
+			splitSize := strings.Split(size, ":")
+			if len(splitSize) != 2 {
+				return fmt.Errorf("Argument to --image-size %s is not "+
+					"in the correct format", size)
+			}
+			parsedSize, err := quantity.ParseSize(splitSize[1])
+			if err != nil {
+				return fmt.Errorf("Failed to parse argument to --image-size: %s",
+					err.Error())
+			}
+			// the image size parsed successfully, now find which volume to associate it with
+			volumeNumber, err := strconv.Atoi(splitSize[0])
+			if err == nil {
+				// argument passed was numeric.
+				if volumeNumber < len(stateMachine.volumeOrder) {
+					stateName := stateMachine.volumeOrder[volumeNumber]
+					stateMachine.imageSizes[stateName] = parsedSize
+				} else {
+					return fmt.Errorf("Volume index %d is out of range", volumeNumber)
+				}
+			} else {
+				if _, found := stateMachine.gadgetInfo.Volumes[splitSize[0]]; !found {
+					return fmt.Errorf("Volume %s does not exist in gadget.yaml",
+						splitSize[0])
+				}
+				stateMachine.imageSizes[splitSize[0]] = parsedSize
+			}
+		}
+	}
+	return nil
+}
+
+// saveVolumeOrder records the order that the volumes appear in gadget.yaml. This is necessary
+// to preserve backwards compatibility of the command line syntax --image-size <volume_number>:<size>
+func (stateMachine *StateMachine) saveVolumeOrder(gadgetYamlContents string) {
```

Comment:
ygh. You handled this best way possible I think, but I still have it super bad. If only snapd could just unmarshal the volumes into something like an ordered map and let us consume it that way - that would be awesome. But I guess changing the ABI like this isn't to be done lightly.